src/python/pants/util/ordered_set.py

```diff
@@ -72,19 +74,28 @@ def __eq__(self, other: Any) -> bool:
             return NotImplemented
         return all(x == y for x, y in itertools.zip_longest(self._items, other._items))
 
-    def union(self, *others: Iterable[T]) -> "_AbstractOrderedSet[T]":
+    def __or__(self: _TAbstractOrderedSet, other: Iterable[T]) -> _TAbstractOrderedSet:  # type: ignore[override]
+        # Differences with AbstractSet: our set union forces "other" to have the same type. That is, while
+        # AbstractSet allows {1, 2, 3} | {(True, False)} resulting in Set[Union[int, Tuple[bool, bool]]], the analogous
+        # for descendants  of _TAbstractOrderedSet is not allowed.
+        # GOTCHA: given higher kinded type _TAbstractOrderedSet[S]:
+        #   if T <: S => _TAbstractOrderedSet[S] => *appears* to perform unification but it doesn't
+        #   if S <: T => type error with this implementation (while AbstractSet would resolve to AbstractSet[T])
+        return self.union(other)
+
+    def union(self: _TAbstractOrderedSet, *others: Iterable[T]) -> _TAbstractOrderedSet:
         """Combines all unique items.
 
         Each item's order is defined by its first appearance.
         """
-        merged_iterables = itertools.chain([self], others)
+        merged_iterables: Iterable[Iterable] = itertools.chain([self], others)
```

Comment:
What's the reason for adding this? If to make MyPy happy, we prefer using `cast`: https://www.pantsbuild.org/docs/style-guide#prefer-cast-to-override-annotations

Also, is it possible to be `Iterable[Iterable[T]]`?

Reply:
Excellent question, thank you for asking. I suspect it's because we annotated `self` and mypy no longer knows that `self` is also an `Iterable[T]` even though `_TAbstractOrderedSet` is bound by `AbstractOrderedSet` which, in turn, is an `Iterable[T]`. I think this boils down to the fact that mypy doesn't fully support higher kinded types (even though we can do some work arounds like this PR). The `Intersection` type is still yet to be implemented so we cannot do anything fancy. Let's look at the following experiment (which shows what I think is the most elegant solution):
```python
from dataclasses import dataclass
from itertools import chain

from typing import TypeVar, Iterator, Iterable, cast

T = TypeVar('T')
TContainer = TypeVar('TContainer', bound='Container')

@dataclass
class Container(Iterable[T]):  # Note that Container is explicitly declared as Iterable[T] but the problem persists
    x: Iterable[T]

    def __iter__(self: TContainer) -> Iterator[T]:
        return iter(self.x)

    def func1(self, *others: Iterable[T]) -> 'Container[T]':
        merged = chain(others, [self])  # : chain[Iterable[T]]
        more_merged = chain.from_iterable(merged)  # : Iterator[T]
        reveal_locals()
        return type(self)(more_merged)

    def func2(self: TContainer, *others: Iterable[T]) -> TContainer:
        merged = chain(others, [self])  # : chain[object]
        more_merged = chain.from_iterable(merged)  # ERRORS (cannot infer type) because of the previous step
        reveal_locals()
        return type(self)(more_merged)

    def func3(self: TContainer, *others: Iterable[T]) -> TContainer:
        merged = chain(others, [cast(Iterable[T], self)])  # : chain[Iterable[T]] => FIXES THE PROBLEM
        more_merged = chain.from_iterable(merged)  # : Iterator[T]
        reveal_locals()
        return type(self)(more_merged)
```
Personal note: It's great to have both structural and nominal subtyping at the same time but life gets complicated if we want to be theoretically rigorous (think about this sort of situations, add nested levels of covariant/contravariant type variables to that)